tachyon/crypto/hashes/sponge/poseidon/poseidon.h

```diff
@@ -44,6 +49,58 @@ struct PoseidonSponge final : public PoseidonSpongeBase<PoseidonSponge<F>> {
     state.elements = math::MulMatVecSerial(config.mds, state.elements);
   }
 
+  void ApplyARKEfficient(SpongeState<F>& state, Eigen::Index round_number,
+                         bool) const {
+    state.elements += config.optimized_ark.row(round_number);
+  }
+
+  void ApplyMixEfficient(SpongeState<F>& state, Eigen::Index index,
+                         bool is_full_round) const {
+    if (is_full_round) {
+      state.elements =
+          math::MulMatVecSerial(config.pre_sparse_mds, state.elements);
```

Comment:
Can you be sure that for all cases the serial version suits better?

Reply:
As you can see in `poseidon_config_entry.h` or `poseidon_transcript.h`, the maximum rate is 8. Additionally, in `matrix_operation_benchmark.cc`, for sizes up to 16, `math::MulMatVecSerial` performs better.